File: quicksort2.php
<?php
/**
 *快速排序优化版本
 */

//块笋插入排序
function insertsort(&$arr, $length)
{
    for ($i = 2; $i <= $length; $i++)
    {
        if ($arr[$i] < $arr[$i - 1])
        {
            $temp = $arr[$i];
            for($j = $i - 1; $temp < $arr[$j]; $j--)
            {
            
                $arr[$j + 1] = $arr[$j];
            }
            $arr[$j + 1] = $temp;
        }
    }
}

function temp(&$arr,$i, $j)
{
    $temp = $arr[$i];
    $arr[$i] = $arr[$j];

    $arr[$j] = $temp;

}

function piviotion(&$arr, $low, $high)
{
    //优化一， 三数取中选取关键字
    $mid = intval( ($low +  $high) / 2);
    if ($arr[$low] > $arr[$high])
    {
        temp($arr, $low, $high);
    }

    if ($arr[$mid] > $arr[$high])
    {
        temp($arr, $mid, $high);
    }

    if ($arr[$low] < $arr[$mid])
    {
        temp($arr, $low, $mid);
    }

    $piviokey = $arr[$low];

    while($low < $high)
    {
        while ($low < $high && $arr[$high] >= $piviokey)
        {
            $high--;  
        }
        //优化二，重新复制，不直接交换
        $arr[$low] = $arr[$high];

        //temp ($arr, $low, $high);

        while ($low < $high && $arr[$low] <= $piviokey)
        {
            $low++;

        }
        $arr[$high] = $arr[$low];
        //temp($arr, $low, $high);
    }

    $arr[$low] = $piviokey;
    return $low;
}

function quicksort(&$arr, $low, $high)
{
    //优化三， 数据少的用直接插入排序
    if ($high > 5)
    {
        //优化四， 用尾递归 代替递归
        while($low < $high)
        {
            $pivio = piviotion($arr, $low, $high); 
            quicksort($arr, $low, $pivio - 1);
            $low = $pivio + 1;
        } 
    
    }else {
        insertsort($arr, $high);
    }
}
$arr = array(
    1 => 9,
    2 => 3,
    3 => 6,
    4 => 2,
    5 => 1,
    6 => 5,
    7 => 4,
    8 => 8,
    9 => 7,
);

quicksort($arr, 1, 9);

echo "<pre>";
print_r($arr);

File: quicksort2.c
#include <stdio.h>
#include <malloc.h>
#define MAXSIZE 10


typedef struct{

    int data[MAXSIZE];
    int length;
}SqList;
//交换i,j位置
void temp(SqList *l, int i, int j){
    int temp = l->data[i];

    l->data[i] = l->data[j];
    l->data[j] = temp;
}

//直接插入排序
void insertsort(SqList *l)
{
    int i,j;

    for (i = 2; i <= l->length; i++)
    {
        if(l->data[i] < l->data[i - 1])
        {
            l->data[0] = l->data[i];

            for (j = i - 1; l->data[0] < l->data[j]; j--)
            {
                l->data[j + 1] = l->data[j];
            }

            l->data[j + 1] = l->data[0];
        }
    }
}


int piviotion(SqList *l, int low, int high)
{
    int mid,piviokey;
    mid = (low + high) / 2;

    if (l->data[low] > l->data[high])
    {
        temp(l, low, high);
    }

    if (l->data[mid] > l->data[high])
    {
        temp(l, mid , high);
    }

    if (l->data[low] < l->data[mid])
    {
        temp(l, low, mid);
    }

    piviokey = l->data[low];

    while (low < high)
    {
        while(low < high && l->data[high] > piviokey)
        {
            high--;
        }

        l->data[low] = l->data[high];

        while (low < high && l->data[low] < piviokey)
        {
            low++;
        }

        l->data[high] = l->data[low];

    }

    l->data[low] = piviokey;

    return low;

    
}


void quicksort(SqList *l, int low, int high)
{

    int pivio;

    if (high > 4)
    {

        while (low < high)
        {
            pivio = piviotion(l, low, high);

            quicksort(l, low, pivio - 1);
            low = pivio + 1;

        }
         
    }else {
    
        insertsort(l);
    }

}




void main()
{
    SqList l;
    SqList *ll;

    ll = &l;

    ll->length = 9;
    ll->data[1] = 2;
    ll->data[2] = 1;
    ll->data[3] = 8;
    ll->data[4] = 6;
    ll->data[5] = 7;
    ll->data[6] = 5;
    ll->data[7] = 3;
    ll->data[8] = 4;
    ll->data[9] = 9;

    
    quicksort(ll, 1, 9);

    int i;
    for (i = 1; i <= 9; i++)
    {
    
        printf("%d\n", ll->data[i]);
    }





}














